src/clients/HubPoolClient.ts

```diff
@@ -224,12 +241,16 @@ export class HubPoolClient extends BaseAbstractClient {
    * @param deposit Deposit event
    * @param returns string L1 token counterpart for Deposit
    */
-  getL1TokenForDeposit(deposit: Pick<DepositWithBlock, "quoteBlockNumber" | "originToken" | "originChainId">): string {
+  getL1TokenForDeposit(
+    deposit:
+      | Pick<v2DepositWithBlock, "originChainId" | "originToken" | "quoteBlockNumber">
+      | Pick<v3DepositWithBlock, "originChainId" | "inputToken" | "quoteBlockNumber">
+  ): string {
     // L1-->L2 token mappings are set via PoolRebalanceRoutes which occur on mainnet,
     // so we use the latest token mapping. This way if a very old deposit is filled, the relayer can use the
     // latest L2 token mapping to find the L1 token counterpart.
-
-    return this.getL1TokenForL2TokenAtBlock(deposit.originToken, deposit.originChainId, deposit.quoteBlockNumber);
+    const inputToken = getDepositInputToken(deposit as DepositWithBlock);
```

Comment:
This is a good call and it is definitely a weakness in the current approach. The issue arises partly due to the fact that `getL1TokenForDeposit()` requires only a partial type, and `getDepositInputToken()` expects a complete type. If the implementation of `getDepositInputToken()` were to change (i.e. if it resolved true or false based on a different member of the `Deposit` type then it could incorrectly make the wrong conclusion. I've implicitly accepted those risks in this PR, but it is a definite risk and if you have a suggestion on how to remove these sharp edges then I'm all ears.

Reply:
I'm putting together some generic type guards to see if we can make typescript give us these guarantees. Will link once I have an example!

Reply:
Since this is now tracked in a separate PR I'll mark this thread as resolved.

PR: https://github.com/across-protocol/sdk-v2/pull/537